Objects/unicodeobject.c

```diff
@@ -2332,6 +2332,220 @@ PyUnicodeWriter_WriteUCS4(PyUnicodeWriter *pub_writer,
+        return -1;
+    }
+    Py_ssize_t len = PyUnicode_GET_LENGTH(unicode);
+
+    // Native ASCII
+    if (PyUnicode_IS_ASCII(unicode)
+        && (requested_formats & PyUnicode_FORMAT_ASCII))
+    {
+        return unicode_export(unicode, view,
+                              len, PyUnicode_1BYTE_DATA(unicode),
+                              1, "B", PyUnicode_FORMAT_ASCII);
+    }
+
+    // Native UCS1
+    int kind = PyUnicode_KIND(unicode);
+    if (kind == PyUnicode_1BYTE_KIND
+        && (requested_formats & PyUnicode_FORMAT_UCS1))
+    {
+        return unicode_export(unicode, view,
+                              len, PyUnicode_1BYTE_DATA(unicode),
+                              1, "B", PyUnicode_FORMAT_UCS1);
+    }
+
+    // Native UCS2
+    if (kind == PyUnicode_2BYTE_KIND
+        && (requested_formats & PyUnicode_FORMAT_UCS2))
+    {
+        return unicode_export(unicode, view,
+                              len, PyUnicode_2BYTE_DATA(unicode),
+                              2, "H", PyUnicode_FORMAT_UCS2);
+    }
+
+    // Convert ASCII or UCS1 to UCS2
+    if (kind == PyUnicode_1BYTE_KIND
+        && requested_formats & PyUnicode_FORMAT_UCS2)
+    {
+        Py_UCS2 *ucs2 = PyMem_Malloc((len + 1) * sizeof(Py_UCS2));
+        if (!ucs2) {
+            PyErr_NoMemory();
+            return -1;
+        }
+
+        _PyUnicode_CONVERT_BYTES(Py_UCS1, Py_UCS2,
+                                 PyUnicode_1BYTE_DATA(unicode),
+                                 PyUnicode_1BYTE_DATA(unicode) + len,
+                                 ucs2);
+        ucs2[len] = 0;
+
+        return unicode_export(unicode, view,
+                              len, ucs2,
+                              2, "H", PyUnicode_FORMAT_UCS2);
+    }
+
+    // Native UCS4
+    if (kind == PyUnicode_4BYTE_KIND
+        && (requested_formats & PyUnicode_FORMAT_UCS4))
+    {
+        return unicode_export(unicode, view,
+                              len, PyUnicode_4BYTE_DATA(unicode),
+                              4, BUFFER_UCS4, PyUnicode_FORMAT_UCS4);
+    }
+
+    // Convert ASCII, UCS1 or UCS2 to UCS4
+    if (requested_formats & PyUnicode_FORMAT_UCS4) {
+        Py_UCS4 *ucs4 = PyUnicode_AsUCS4Copy(unicode);
+        if (ucs4 == NULL) {
+            return -1;
+        }
+        return unicode_export(unicode, view,
+                              len, ucs4,
+                              4, BUFFER_UCS4, PyUnicode_FORMAT_UCS4);
+    }
+
+    // Encode UCS1, UCS2 or UCS4 to UTF-8
+    if (requested_formats & PyUnicode_FORMAT_UTF8) {
+        Py_ssize_t nbytes;
+        const char *utf8 = PyUnicode_AsUTF8AndSize(unicode, &nbytes);
+        if (utf8 == NULL) {
+            return -1;
+        }
+        return unicode_export(unicode, view,
+                              nbytes, utf8,
+                              1, "B", PyUnicode_FORMAT_UTF8);
+    }
+
+    PyErr_Format(PyExc_ValueError, "unable to find a matching export format");
```

Comment:
```suggestion
    PyErr_Format(PyExc_ValueError, "unable to find a matching export format in: %" PRIu32, requested_formats);
```

Strictly speaking `uint32_t` requires to use `PRIu32` from `inttypes.h` as a format specifier, but maybe the `%i` is sufficient.

Reply:
I don't think that formatting the format value will help users, it's a bit mask.

Reply:
Ah yes, then I'd use `PyErr_SetString` instead.

Reply:
Fixed.